Mage/src/main/java/mage/abilities/effects/common/PutRandomCardFromLibraryIntoGraveyardEffect.java

```diff
@@ -0,0 +1,65 @@
+package mage.abilities.effects.common;
+
+import mage.abilities.Ability;
+import mage.abilities.effects.OneShotEffect;
+import mage.cards.Card;
+import mage.constants.Outcome;
+import mage.constants.Zone;
+import mage.filter.FilterCard;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.util.RandomUtil;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * @author karapuzz14
+ */
+public class PutRandomCardFromLibraryIntoGraveyardEffect extends OneShotEffect {
+    private FilterCard filter;
+
+    /**
+     * @param filter for selecting a card
+     */
+    public PutRandomCardFromLibraryIntoGraveyardEffect(FilterCard filter) {
+        super(Outcome.Discard);
+        this.filter = filter;
+        this.staticText = "put a random " + filter.getMessage() + " from your library into your graveyard";
+    }
+
+    private PutRandomCardFromLibraryIntoGraveyardEffect(final PutRandomCardFromLibraryIntoGraveyardEffect effect) {
+        super(effect);
+        this.filter = effect.filter;
+    }
+
+    @Override
+    public PutRandomCardFromLibraryIntoGraveyardEffect copy() {
+        return new PutRandomCardFromLibraryIntoGraveyardEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        Permanent sourcePermanent = game.getPermanentOrLKIBattlefield(source.getSourceId());
```

Comment:
sourcePermanent isn't used anywhere?

Reply:
> sourcePermanent isn't used anywhere?

It's assumed that sourcePermanent is used for confirmation of source permanent existence (at line 45).

Reply:
Not necessary. The permanent doesn't have to still be on the battlefield for the effect to resolve, which is why it's checking LKI, but the LKI isn't relevant to anything about the effect either.

Reply:
Okay. Then I should remove it.